Detect home server from matrix id

Kaki-In opened the pull request:
# Detect home server from matrix id

## Added functionnality : 
 - you can now enter the user matrix id (e.g. @username:server.local) in the login input. 
 - if the home server that you've entered into youre id is not the same that those into the home server input, it will be use instead to connect to the default matrix server.

Should fix [this issue (#479)](https://github.com/vector-im/hydrogen-web/issues/479).

Kaki-In committed:
Adding the functionnality to use his own matrix id (@username:server.host) in the username input when logging in, without having to update the server hostname on the first input

Kaki-In committed:
Added some forgot changes

Kaki-In committed:
Removed a console.log

Kaki-In committed:
Added some forgot changes

Kaki-In committed:
Removed a console.dir

Kaki-In committed:
Fixed minor bug

Kaki-In committed:
Fixed the problem of having given a bad password/username

Kaki-In committed:
Improve the connection frame actualisation

Kaki-In committed:
Fixed a bug that always showed the password and login views when loading the messages list

bwindels commented on src/domain/login/PasswordLoginViewModel.js:
```diff
@@ -39,10 +41,35 @@ export class PasswordLoginViewModel extends ViewModel {
+    async parseUsernameLogin (login) {
+        if ((login.match(/@/g) || []).length == 1 && (login.match(/:/g) || []).length == 1 && (login.match(/\./g) || []).length) {
+            if (login.indexOf("@") == 0 && login.indexOf("@") < login.indexOf(":") && login.indexOf(":") < login.indexOf(".")) {
+				if (this._loginOptions.homeserver != login.substring(login.indexOf(":") + 1)) {await await this._options.setHomeserver(login.substring(login.indexOf(":") + 1));}
```
You added the double await because you're now returning a promise from an async method?

Kaki-In replied:
Yes, exactly. I am not sure to have to put the await inside the `setHomeServer` method.

MidhunSureshR replied:
When you do `const foo = await something...`, foo will never be a promise!
See https://jsfiddle.net/e3tpmufg/

bwindels commented on src/domain/login/PasswordLoginViewModel.js:
```diff
@@ -39,10 +41,35 @@ export class PasswordLoginViewModel extends ViewModel {
+	}
 
+    async parseUsernameLogin (login) {
+        if ((login.match(/@/g) || []).length == 1 && (login.match(/:/g) || []).length == 1 && (login.match(/\./g) || []).length) {
```
Consider using just one regex to parse the mxid and extract the server name : `@.*:(.+)`

bwindels reviewed (commented):
Just a few comments, haven't done a full review yet.

Kaki-In committed:
New corrections (from base of #810 pull request commentaries) :
 - fix methods miss;
 - corrected some identation mistakes.

Kaki-In committed:
Update about the serverName constant : missing the result id (from the regex)

Kaki-In committed:
Detected inattention mistake.
PasswordLoginViewModel.parseUsernameLogin() should return the username, not the servername.

Kaki-In committed:
Merge branch 'vector-im:master' into detect-hs-from-mxid

MidhunSureshR commented on src/platform/web/ui/login/PasswordLoginView.js:
```diff
@@ -32,15 +32,15 @@ export class PasswordLoginView extends TemplateView {
         });
         
-        return t.div({className: "PasswordLoginView form"}, [
+        return t.div({className: "PasswordLoginView form" + (disabled ? " disabled" : "")}, [
```
You can do this without string manipulation:
```suggestion
return t.div({
	className: {
		PasswordLoginView: true,
		form: true,
		disabled
	}
}, [
```

MidhunSureshR commented on src/platform/web/ui/login/PasswordLoginView.js:
```diff
@@ -18,7 +18,7 @@ import {TemplateView} from "../general/TemplateView";
 export class PasswordLoginView extends TemplateView {
     render(t, vm) {
-        const disabled = vm => !!vm.isBusy;
+        const disabled = vm => (!!vm.isBusy) && (!vm.isEnabled);
```
Not a change you made but we don't really need the double bang (`!!`) here since `vm.isBusy` is already a boolean.
So this can be:
```suggestion
        const disabled = vm => vm.isBusy && !vm.isEnabled;
```

MidhunSureshR commented on src/domain/login/PasswordLoginViewModel.js:
```diff
@@ -39,10 +41,36 @@ export class PasswordLoginViewModel extends ViewModel {
+        this.emitChange("isEnabled");
+    }
 
+    async parseUsernameLogin (login) {
```
```suggestion
    async parseUsernameLogin (username) {
```

MidhunSureshR replied:
You can do something like:
```js
findHomeserverFromUsername (login) {
    const matches = login.match(/@.*:(.+)/);
    if (!matches) {
        return;
    }
    const [, homeserverInMxid] = matches;
    return homeserverInMxid;
}

async changeHomeserverFromUsername(username) {
    const currentHomeserver = this._loginOptions.homeserver;
    const newHomeserver = this.findHomeserverFromUsername(username);
    if (newHomeserver && currentHomeserver !== newHomeserver) {
        await this._options.setHomeserver(newHomeserver);
        return true;
    }
    return false;
}
```

MidhunSureshR commented on src/domain/login/PasswordLoginViewModel.js:
```diff
@@ -57,7 +85,13 @@ export class PasswordLoginViewModel extends ViewModel {
-            this._showError(error);
+            if (usernameChanged) {
+                console.log(error);
+                this._options.showConnectionError(error);
```
Why do we have two different ways of showing errors here?

Kaki-In replied:
Surely a miss from a previous commit

MidhunSureshR reviewed (commented):
Some comments and questions:

Kaki-In committed:
Resolved the comments at https://github.com/vector-im/hydrogen-web/pull/810
Fix all the new comments inside https://github.com/vector-im/hydrogen-web/pull/810#pullrequestreview-1050539767

bwindels commented on src/platform/web/ui/login/PasswordLoginView.js:
```diff
@@ -18,7 +18,7 @@ import {TemplateView} from "../general/TemplateView";
 export class PasswordLoginView extends TemplateView {
     render(t, vm) {
-        const disabled = vm => !!vm.isBusy;
+        const disabled = vm => vm.isBusy && vm.isEnabled;
```
This seems weird. We disable the form when we're both busy and the view model says we should be enabled rather than disabled? Do we even need the two fields?

Kaki-In replied:
Yes, it should be `vm.isBusy || !vm.isEnabled;` sorry ^^'

bwindels commented on src/matrix/common.js:
```diff
@@ -22,16 +22,16 @@ export function makeTxnId() {
 }
 
 export function tests() {
-	return {
```
Please don't change the indenting when not making changes.

Kaki-In committed:
Fixed an error while writing boolean

Kaki-In committed:
Canceled the indenting modification.

Kaki-In committed:
Merge branch 'master' into detect-hs-from-mxid

Kaki-In committed:
Resolved morge conflict with LoginViewmodel

Kaki-In committed:
Merge branch 'vector-im:master' into detect-hs-from-mxid

MidhunSureshR commented on src/domain/login/LoginViewModel.ts:
```diff
@@ -193,20 +206,17 @@ export class LoginViewModel extends ViewModel<Options> {
-        this._passwordLoginViewModel = this.disposeTracked(this._passwordLoginViewModel);
-        this._completeSSOLoginViewModel = this.disposeTracked(this._completeSSOLoginViewModel);
-        this.emitChange("disposeViewModels");
+    updateLoginOptions(): void {
```
```suggestion
    private _updateLoginOptions(): void {
```
Don't forget to add the "_" to where this method is called.

MidhunSureshR commented on src/domain/login/LoginViewModel.ts:
```diff
@@ -123,26 +124,38 @@ export class LoginViewModel extends ViewModel<Options> {
-                attemptLogin: (loginMethod: PasswordLoginMethod) => this.attemptLogin(loginMethod)
-        })));
+                attemptLogin: loginMethod => this.attemptLogin(loginMethod),
+                setHomeserver: homeserver => this.setHomeserver(homeserver)
```
```suggestion
                attemptLogin: (loginMethod: PasswordLoginMethod) => this.attemptLogin(loginMethod),
                setHomeserver: (homeserver: string) => this.setHomeserver(homeserver)
```

MidhunSureshR reviewed (commented):
Some more ts changes:

Kaki-In committed:
Fixed last checked on https://github.com/vector-im/hydrogen-web/pull/810

Kaki-In committed:
Merge branch 'vector-im:master' into detect-hs-from-mxid

Kaki-In committed:
Merge branch 'vector-im:master' into detect-hs-from-mxid
